daemons/based/based_callbacks.c

```diff
@@ -1312,6 +1313,7 @@ cib_process_command(xmlNode * request, xmlNode ** reply, xmlNode ** cib_diff, gb
 
             } else if (pcmk__str_eq(section, XML_CIB_TAG_CONFIGURATION, pcmk__str_casei)) {
                 send_r_notify = TRUE;
+                skip_election = TRUE;
```

Comment:
Hi Ken,

I was also interested in the movement and intended to confirm it.
Let's check it again.

Many thanks,
Hideo Yamauchi.

Reply:
Hi Ken,

It's a change of the node section, but is there a problem just to confirm that the diff operation is the node section obediently?

```
 <change operation="delete" path="/cib/configuration/nodes/......................."/>
```

Best Regards,
Hideo Yamauchi.

Reply:
Hi Ken,

Thanks for your comment.

> Currently, the only daemons that subscribe to T_CIB_REPLACE_NOTIFY notifications are attrd and controld. Both need to know whether the nodes changed (in configuration or status). However, attrd (via attrd_cib_replaced_cb()) also needs to know whether the alerts section changed. The current cib_process_command() code will send a notification if the entire cib changes, or if the configuration section changes, but it will not send a notification if only the alerts section changes (there is no test for XML_CIB_TAG_ALERTS). Thus, attrd will not change its behavior if the user modifies a node attribute alert by replacing the alerts section. The reproducer would be cibadmin --replace.
> 

Okay.
I forgot that I could specify alert for the scope of the cibadmin command.

> Looking at it again, the current code also does not send notifications for a single node, alert, or alert recipient changing (XML_CIB_TAG_NODE, XML_CIB_TAG_STATE, XML_CIB_TAG_ALERT, or XML_CIB_TAG_ALERT_RECIPIENT). Checking for replacement of only an instance_attributes, meta_attributes, or select element that happens to be inside an alert would be more difficult.
> 
> I'm thinking a better approach for this part of the code might be to hash the nodes, alerts, and status sections, and compare only those hashes, instead of comparing _section_ against specific element names.

Does it mean something like the following?

```
(Details are omitted)
(snip)
        cib_nodes = get_xpath_object("//"XML_CIB_TAG_CONFIGURATION"/"XML_CIB_TAG_NODES, current_cib, LOG_TRACE);
        current_node_digest = calculate_xml_versioned_digest(cib_nodes, FALSE, TRUE, CRM_FEATURE_SET);
        ....
        cib_nodes = get_xpath_object("//"XML_CIB_TAG_ALERT, current_cib, LOG_TRACE);
        current_alert_digest = calculate_xml_versioned_digest(cib_nodes, FALSE, TRUE, CRM_FEATURE_SET);
        ....
        cib_nodes = get_xpath_object("//"XML_CIB_TAG_STATUS, current_cib, LOG_TRACE);
        current_status_digest = calculate_xml_versioned_digest(cib_nodes, FALSE, TRUE, CRM_FEATURE_SET);
        ....
(snip)
        if (pcmk__str_eq(CIB_OP_REPLACE, op, pcmk__str_none)) {
            gboolean change_node = FALSE;
            gboolean change_alert = FALSE;
            gboolean change_status= FALSE;

            cib_nodes = get_xpath_object("//"XML_CIB_TAG_CONFIGURATION"/"XML_CIB_TAG_NODES, result_cib, LOG_TRACE);
            result_node_digest = calculate_xml_versioned_digest(cib_nodes, FALSE, TRUE, CRM_FEATURE_SET);
            ....
            cib_nodes = get_xpath_object("//"XML_CIB_TAG_ALERT, result_cib, LOG_TRACE);
            result_alert_digest = calculate_xml_versioned_digest(cib_nodes, FALSE, TRUE, CRM_FEATURE_SET);
            ....
            cib_nodes = get_xpath_object("//"XML_CIB_TAG_STATUS, result_cib, LOG_TRACE);
            result_status_digest = calculate_xml_versioned_digest(cib_nodes, FALSE, TRUE, CRM_FEATURE_SET);
            ....
            //Compare hashes (digests)
            if (pcmk__str_eq(current_node_digest, result_node_digest, pcmk__str_casei)) {
                change_node = TRUE;
            }
            if (pcmk__str_eq(current_alert_digest, result_alert_digest, pcmk__str_casei)) {
                change_alert = TRUE;
            }
            if (pcmk__str_eq(current_status_digest, result_status_digest, pcmk__str_casei)) {
                change_status = TRUE;
            }
            
            if (change_node || change_alert || change_status) {
                send_r_notify = TRUE;
            }
            if (!change_node && !change_status) {
                skip_election = TRUE;
            }
(snip)
```


Many thanks,
Hideo Yamauchi.

Reply:
> Does it mean something like the following?

Hi Hideo,

Yes, that's what I had in mind. The digest function can't handle NULL inputs so you'll have to skip those, and we probably want to look at the whole XML_CIB_TAG_ALERTS section.

Reply:
> Yes, that's what I had in mind. The digest function can't handle NULL inputs so you'll have to skip those, and we probably want to look at the whole XML_CIB_TAG_ALERTS section.

Hi Ken,

Okay!
Calculates the entire section of XML_CIB_TAG_ALERTS.

I'm trying to sort in the xml section considering the replacement in the section, but at present, it seems that I can't find a function that sorts everything in the xml section.
Do you know a function that sorts everything in a section?


Many thanks,
Hideo Yamauchi.

Reply:
Hi Hideo, see sorted_xml()

Reply:
> Hi Hideo, see sorted_xml()

Okay!
Thanks!

Reply:
The sorted_xml () seems to sort the nvpair in the section.
But I think I need to sort by id in the section first, shouldn't I consider it?

```
This is an example of the node section, but in the xml to be updated, the order is just changed, but the contents are the same.
I would like to judge this as the same section.

(current)
    <nodes>
      <node id="3232238190" uname="srv02"/>
      <node id="3232238180" uname="srv01"/>
    </nodes>

(change)
    <nodes>
      <node id="3232238180" uname="srv01"/>
      <node id="3232238190" uname="srv02"/>
    </nodes>
```

Reply:
Oh right -- sorted_xml() only sorts the attributes of each element, it doesn't sort the elements themselves. I don't think we have anything that does that. It might not be worth the effort to catch order-only changes.

Reply:
Hi Ken,

> Oh right -- sorted_xml() only sorts the attributes of each element, it doesn't sort the elements themselves. I don't think we have anything that does that. It might not be worth the effort to catch order-only changes.

I think so, too.
For now, let's ignore this change in order.

Many thanks,
Hideo Yamauchi.